mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,164 @@
+# Authors: George O'Neill <g.o'neill@ucl.ac.uk>
+#
+# License: BSD-3-Clause
+
+from numpy import linalg, arange, isnan, zeros, eye, shape, nan_to_num
+from numpy import any as npany
+from mne.preprocessing.maxwell import _prep_mf_coils, _sss_basis
+from mne.io.pick import pick_types
+from ..utils import _check_preload, warn
+from ..rank import estimate_rank
+
+
+# @verbose
+def apply_hfc(raw, order=1, exclude_bads=True, verbose=None):
+    """Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous or harmonic field. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The data instance to process.
+    order: int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    exclude_bads: bool
+        Do not include bad channels in the projection, or alter their data
+        (default: True).
+    %(verbose)s
+
+    Returns
+    -------
+    proc : Raw
+        The processed data.
+    """
+
+    if exclude_bads is True:
+        picks_idx = pick_types(raw.info, meg=True)
+    else:
+        picks_idx = pick_types(raw.info, meg=True, exclude=None)
+    sens_idx = _filter_channels_with_positions(raw.info, picks_idx)
+    model = FieldCorrector(raw, picks=sens_idx, order=order)
+    proc = model.apply(raw)
+
+    return proc
+
+
+# @fill_doc
+class FieldCorrector():
```

Comment:
Agreed.

Reply:
@georgeoneill thinking about this again, these just end up as projection vectors, right? If so I wonder if we should instead expose it as `mne.preprocessing.compute_hfc_proj` or something, then it can be applied like any other projection vector (e.g., empty-room projector, blink projector, ECG proj, etc.). Then we get some nice stuff that goes along with it (viz, incorporation into inverses properly, etc.).

Reply:
This is something we can consider. However, I have found from my inital playing with the projector systems that the two don't always give the same results. Something to do with the normalisation of the vectors I think? So I would have rather allow both the online projectors and offline correction to be available to the end user.

Reply:
> I have found from my inital playing with the projector systems that the two don't always give the same results. Something to do with the normalisation of the vectors I think?

Yeah it's probably because you multiply by the `pinv` of the un-normalized vectors here, rather than ortho-normalizing them ahead of time, which is what we do in the proj code.

One motivation is that, looking at the code here, you do a lot of picks wrangling, I/O stuff, linear algebra, etc. This is all taken care of if we just treat these N harmonic components as N projection vectors instead. I think this code would be ~100 lines instead of ~500, which helps with maintainability.

Another motivation is that I worry that this current code won't always be accounted for properly in inverse modeling. When you use `proj` we know you're reduced your rank by exactly `N` components and this is taken into account during inverse modeling so you don't accidentally amplify the zeroed out components during whitening. Without orthonormalizing the bases, IIRC your N components will not all be fully projected out to the same ~1e-14 numerical noise level (but probably close / within a couple orders of magnitude?), won't show up in `info['proj']` etc. We face a similar problem with ICA actually and inversion seems to be okay there (maybe with regularization?), so maybe it's not a big problem in practice, but that remains to be seen/checked. If we use `proj`, we know we're safe here I think.

Maybe an intermediate way to go would be to see how far we can get using `proj` code for I/O and viz etc. that hopefully doesn't require normalization (I don't think it does?). Even if we want the `pinv`-based method it will still cut down on code dup. In other words, 

1. Let your computation function return a list of un-normalized projection vectors, and use this for I/O and viz etc. 2. Add a private (for now) function that takes `N` projection vectors and -- without normalizing them -- creates a projection matrix based on `_make_projector` / `eye(n_ch) - S @ pinv(S)` like you have here.
3. Use the private function to check against the MATLAB implementation for numerical similarity.
4. Use the standard `proj` stuff in public code/tutorials (for now).

That would allow us to keep code slim and punt on (publicly) doing projections to the data that do not go through `info['projs']` / normalization. WDYT?

If you try (1) and immediately run into problems where things must be normalized (e.g., in I/O), we could see if it would be easy enough to modify the proj I/O code to support un-normalized vectors. I'm happy to help with some or all of this as needed, too!

Reply:
Sorry for the radio silence, just acknowleding this now. Will report back when I have something which attempts to follow this.

Reply:
Sounds good!

Reply:
To no-ones suprise (other than myself) it appears to work, whilst still passing all the unit tests with no modification! Top panel is with the projectors turned off, bottom panel with them turned on. 

![image](https://user-images.githubusercontent.com/3579812/231576808-ed6d2d37-b9f7-436d-9f59-4e1684ab21c2.png)

Commit coming imminently.

Reply:
Great! If we can indeed use the projector interface, we don't need a fit/apply or class at all. I suggest we just have `mne.preprocessing.compute_proj_hfc` that returns a list of `Projector` instances. Then they can be treated as any other projector in terms of `raw.add_proj(projs).apply_proj()`, delayed projection, I/O, viz, reconstruction, ...

Reply:
i.e., it could have an interface that would immediately blend with and benefit from all the stuff in https://mne.tools/dev/auto_tutorials/preprocessing/50_artifact_correction_ssp.html

Reply:
I think I have that working now.